test/dist_autograd_test.py

```diff
@@ -149,10 +154,8 @@ def test_autograd_functions(self):
             # We should have send/recv functions from the previous rank, get all
             # contexts in this node to find them.
 
-            # Wait for the prev rank to be done with rpc.
-            while not prev_rank_rpc_done:
-                time.sleep(0.1)
-                pass
+            # Synchronize all workers.
+            rpc.sync_rpc()
```

Comment:
Lets use `dist.barrier()` since `sync_rpc` might go away in a few future changes we have planned.

Reply:
I don't think `dist.barrier()` works here (just tried it and the tests fail if we replace it). It appears that `dist.barrier()` blocks until all the processes enter the barrier. But `rpc.sync_rpc()` gives us the further guarantee that all outstanding RPCs will be completed, which is what we need here, since we need the RPCs for releasing the context to finish. With `dist.barrier()` there is no guarantee that all outstanding RPCs will be completed, it looks like. cc @mrshenli in case my understanding isn't correct

Reply:
Hmm what is the reason `dist.barrier` doesn't work here? We are only doing sync RPCs in this test so all RPCs should have finished on all nodes when we are done with the barrier.

> Why is sync_rpc going away? Is there going to be any other way to enforce all processes finish all outstanding RPCs and then move forward (without joining)?

Because the user should wait on RPCs that they care about and not the framework. If you care about an RPC, you should use future.wait() to check its result.

Reply:
But the RPC that is sent to release the context on other nodes isn't synchronous right? It is sent with `agent->send()` which enqueues the send, and the user doesn't wait on this (i.e. when the context manager is exited, the RPCs are sent, but we don't wait on these RPCs). If we want, we can wait on the future inside `releaseContext`, but that would be a slowdown to the user.

Reply:
In this particular case `dist.barrier()` works since the async RPCs haven't been sent, but in the other test where we sync after exiting the context manager, we would need a way to make sure the RPCs are sent + processed before continuing with the test.

Reply:
> Because the user should wait on RPCs that they care about and not the framework.

Agreed. What about the case where our framework itself sends async RPCs that as part of handling a user request? This is basically what these cleanup context RPCs are. The user does not have any handle to the futures created by them, but we should at some point guarantee that they are done.